R/compute_metrics.R

```diff
@@ -0,0 +1,160 @@
+#' metrics are summarized.
+#' @inheritParams control_grid
+#' @inheritParams collect_metrics
+#'
+#' @return A tibble. See [collect_metrics()] for more details on the return value.
+#'
+#' @examplesIf tune:::should_run_examples()
+#' # load needed packages:
+#' library(parsnip)
+#' library(rsample)
+#' library(yardstick)
+#'
+#' # evaluate a linear regression against resamples.
+#' # note that we pass `save_pred = TRUE`:
+#' res <-
+#'   fit_resamples(
+#'     linear_reg(),
+#'     mpg ~ cyl + hp,
+#'     bootstraps(mtcars, 5),
+#'     control = control_grid(save_pred = TRUE)
+#'   )
+#'
+#' # to return the metrics supplied to `fit_resamples()`:
+#' collect_metrics(res)
+#'
+#' # to compute new metrics:
+#' compute_metrics(res, metric_set(mae))
+#'
+#' # if `metrics` is the same as that passed to `fit_resamples()`,
+#' # then `collect_metrics()` and `compute_metrics()` give the same
+#' # output, though `compute_metrics()` is quite a bit slower:
+#' all.equal(
+#'   collect_metrics(res),
+#'   compute_metrics(res, metric_set(rmse, rsq))
+#' )
+#' @name compute_metrics
+#' @export
+compute_metrics <- function(x, metrics, summarize, event_level, ...) {
+  UseMethod("compute_metrics")
+}
+
+#' @export
+#' @rdname compute_metrics
+compute_metrics.default <- function(x, metrics, summarize, event_level, ...) {
+  rlang::abort("No `compute_metrics()` method exists for this type of object.")
+}
+
+#' @export
+#' @rdname compute_metrics
+compute_metrics.tune_results <- function(x,
+                                         metrics,
+                                         summarize = TRUE,
+                                         event_level = "first",
+                                         ...) {
+  if (!any(names(x) == ".predictions")) {
+    rlang::abort(paste0(
+      "The inputted `x` argument must have been generated with the ",
+      "control argument `save_pred = TRUE`."
+    ))
+  }
+
+  if (!inherits(metrics, "metric_set")) {
+    rlang::abort("The inputted `metrics` argument must be a metric set.")
+  }
+
+  new_metrics_info <- metrics_info(metrics)
+  old_metrics_info <- metrics_info(attr(x, "metrics"))
+
+  # if only a class metric is supplied to a tuning function, then
+  # `save_pred = TRUE` will only return the class predictions from
+  # each model. same story goes for prob metrics.
+  if (!all(new_metrics_info$type %in% old_metrics_info$type)) {
+    cli::cli_abort(c(
+      "The supplied `metrics` argument has input  \\
+       {cli::qty(new_metrics_info$type)} type{?s} {.val {new_metrics_info$type}}, \\
+       while the metrics used to generate predictions only used \\
+       {.val {old_metrics_info$type}} metrics.",
+      "i" = "To save predictions for {new_metrics_info$type} metrics, \\
+             generate {.arg x} with metrics of that type."
+    ))
+  }
+
+  if (!inherits(summarize, "logical") || length(summarize) != 1L) {
+    rlang::abort("The `summarize` argument must be a single logical value.")
+  }
```

Comment:
Do you use type checkers here yet? This is just `check_bool(summarize)`

Reply:
Not yet, but I agree this ought to be transitioned over when that happens.